### DISCUSSION TITLE
WARN No native ExponentConstants module found

### DISCUSSION THREAD

**Discussion by Lalman888** [expo/expo#24595] [Archived: Questions]
1

I've updated the expo version in my react native project from "~43.0.2" to "^48.0.4" and then the error from the title appears. The 43.0.2 version works perfectly, but the 48.0.4 version fails to work, showing me the next error:

WARN No native ExponentConstants module found, are you sure the expo-constants's module is linked properly?

ERROR Error: Expo native runtime is not available: something went wrong and we aren't sure what it was. Please post more information and get support at https://forums.expo.dev/.

ERROR Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect. This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native.

ERROR Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect. This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native.

**Comment by tsapeta** [comment]
Could you provide some more details, e.g. which packages and versions do you use? Can you try running `npx expo install --fix`?

**Comment by pbinmich** [comment]
Upgrade your expo-modules-core (I did to version 1.5.12). The expo-constants package has ExponentConstants.ts that references 'requireNativeModule.ts' in expo-modules-core which doesn't exist in earlier versions. The latest expo (49 at this time) specs expo-modules-core 1.5.12 but it wasn't installed - somehow I had a REALLY old version (0.2.0).

yarn add expo-modules-core@1.5.12

**Reply by tsapeta** [reply]
Installing `expo-modules-core` yourself is never a good idea. I think it was just broken because you had it (v0.2.0) installed directly in your app. It should only be a transitive dependency of the `expo` package for the proper compatibility.

Can you please try to remove it from your dependencies and see if the issue is fixed?

**Reply by pbinmich** [reply]
Thanks for the feedback. I'm very new to the React Native world, it's been 20+ years since I touched Javascript and web apps. I've been handed the task of upgrading an old React Native app from 0.63.3 - during this process I came across this exact same error. Initially, I just tried upgrading incremental versions (of RN) - but then switched to just installing the latest versions of RN, Expo, tools, etc., and started with a clean slate (init'd empty managed Expo app), moving the code over and handling missing dependencies as reported (so I started with an empty packages.json). As far as I understand, this would be the cleanest, most up to date, starting point.

My point is, I never installed expo-modules-core directly (initially). From all the latest versions of Expo packages, somehow it resulted in having version 0.2.0 - which is ancient (100+ versions back). This *was* the root of my error, the missing 'requireNativeModule' from that package - it seems to appear around version 0.4.0. I'd have thought all the new versions of packages would have resulted in a newer version of expo-modules-core, especially if I'm not expected to install this manually.

I'd follow your advise - but I'm unsure why I ended up with such an old version to begin with. I tried deleting the node_modules dir and did a fresh 'yarn install' after *removing* the expo-modules-core from my package.json. It resulted in the same 0.2.0 version of expo-modules-core.

I'm a bit confused - the 'expo' package itself lists the dependency of expo-modules-core at 1.5.12. So I'm unsure why 0.2.0 gets installed. I'd like to understand this better - although at this time, making sure the right version is installed (for a clean install) requires me to list expo-modules-core in my app package.json at version 1.5.12. :(

Hoping this explains why I was forced to do this. I would expect others to experience the same (upgrading apps?). Could this be from other package dependencies being incompatible with a newer version, forcing the install of a much older one? If so, I'd prefer the build to fail, not deal with a vague runtime error (problem being the dependency of a dependency). Maybe I'm missing something respective to how this package management works (again, I'm very new to all this).

Feedback greatly appreciated.

UPDATE:

'yarn why expo-modules-core' 

=> Found "expo-modules-core@0.2.0"
info Has been hoisted to "expo-modules-core"
info Reasons this module exists
   - Hoisted from "react-native-unimodules#expo-modules-core"
   - Hoisted from "react-native-unimodules#expo-constants#expo-modules-core"
   - Hoisted from "react-native-unimodules#expo-file-system#expo-modules-core"
...
=> Found "expo#expo-modules-core@1.5.12"
info This module exists because "expo" depends on it.

I have two versions:
node_modules/expo-modules-core (0.2.0)
node_modules/expo/node_modules/expo-modules-core (1.5.12)

So - my app is using react-native-unimodules declaring the 0.2.0 dependency, my yarn.lock file also lists expo-constants having it too (not sure why 'yarn why' doesn't list that one).

Maybe react-native-unimodules is something I need to port my app away from?


UPDATE #2 (!)

FIXED! The application had one reference to react-native-unimodules, respective to app versions. Similar with a couple of references to expo-constants elsewhere. Replaced the code using expo-application to get versions. I removed both react-native-unimodules, expo-modules-core, and expo-constants from the apps package.json file and reinstalled all the packages. Now the expo-modules-core is at the correct version.

Lessons learned! Of course you were quite correct :)

**Reply by tsapeta** [reply]
Nice, I'm happy you were able to solve the issue 😉 react-native-unimodules has been deprecated a long time ago, so I guess your project was created even earlier. You can also try `npx expo-doctor` next time, it should tell you about some issues (not all, though).

**Comment by takunda2013** [comment]
l am also facing this error with React Native Cli after installing this package '@tensorflow/tfjs'
